### PR TITLE
Add Dockerfile for building and Compose project for running

### PR DESCRIPTION
**Pull request by alanking**
The rest of the work for #66 (I think).

Still playing around with the best way to get the configuration into the box and the best way to stand up the services inside the container, but throwing this up here for eyeballs. This *works* but may not be the most user-friendly thing as of yet. README changes will be needed as well, of course.

### CODE REVIEW

File: docker/builder/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+RUN apt-get update && \
+    apt-get install -y \
+        irods-dev \
+        irods-runtime \
+        'irods-externals-boost*' \
+        'irods-externals-json*' \
+        'irods-externals-cmake*' \
+        'irods-externals-clang*' \
+        'irods-externals-pistache*' \
+        'irods-externals-spdlog*' \
```
Comment (alanking) [inline docker/builder/Dockerfile:57]: Update these to be more targeted

File: docker/builder/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        python3-jsonschema \
+        python3-packaging \
+        python3-psutil \
+        python3-pyodbc \
+        python3-requests \
+        sudo \
+        super \
+        wget \
+    && \
+    #pip3 --no-cache-dir install lief && \
```
Comment (alanking) [inline docker/builder/Dockerfile:39]: Probably don't need all of these packages, but that remains to be seen

File: docker/test/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+            dockerfile: nginx.Dockerfile
+        ports:
+            - "80:80"
+
+    irods-client-rest-cpp:
+        build:
+            context: .
+            dockerfile: irods_client_rest_cpp.Dockerfile
+        volumes:
+            - /home/alanking/hdd/builds/irods_client_packages/irods_client_rest_cpp/main/ubuntu-20.04:/packages:ro
```
Comment (alanking) [inline docker/test/docker-compose.yml:16]: D'oh, forgot to remove my home directory here.

This is a provision until we release officially. Until then, we have to provide our own built packages when this is run. Alternatively, the built packages could/should be provided when the Dockerfile is built.

File: docker/test/nginx.Dockerfile [hunk 4]
```diff
@@ -0,0 +1,3 @@
+FROM nginx
```
Comment (alanking) [inline docker/test/nginx.Dockerfile:1]: What version should we pin here?

File: docker/test/irods_client_rest_cpp.json [hunk 5]
```diff
@@ -0,0 +1,85 @@
+{
+    "irods_client_environment": {
+        "zone": "tempZone",
+        "host": "irods-catalog-provider",
+        "port": 1247,
+        "ssl_negotiation_policy": "CS_NEG_REFUSE",
+        "rodsadmin_username": "rods",
+        "rodsadmin_password": "rods"
```
Comment (alanking) [inline docker/test/irods_client_rest_cpp.json:8]: Obviously, this whole section will have to change depending on where the iRODS server is. This provides a strong case for using a volume mount to inject this file when the service is stood up.

File: docker/test/nginx.Dockerfile (hunk 4 above)
Comment (korydraughn) [reply docker/test/nginx.Dockerfile:1]: You could use the latest working version provided by docker hub. That can be our baseline.

Comment (alanking) [reply docker/test/nginx.Dockerfile:1]: Makes sense. I have now tagged this as "1.23.1" as that is in line with "latest" at the time of writing. Unless I should just use "latest".

File: docker/test/docker-compose.yml (hunk 3 above)
Comment (alanking) [reply docker/test/docker-compose.yml:16]: I decided that we should install the local package at image build time. The one building the container must copy the package to the local build context, but I think this is much better than installing when the container starts. This also gives the benefit of being able to restart the REST service (which means we can update the configuration file in the mount point and restart the service to apply the changes).

File: docker/test/nginx.Dockerfile (hunk 4 above)
Comment (korydraughn) [reply docker/test/nginx.Dockerfile]: Using the actual version number like you've done here is what I meant.

You can mark this as resolved.

File: docker/builder/Dockerfile (hunk 1 above)
Comment (alanking) [reply docker/builder/Dockerfile:57]: I think we can optimize this later...?

File: docker/builder/Dockerfile (hunk 2 above)
Comment (alanking) [reply docker/builder/Dockerfile:39]: Same here - can optimize later?

**Comment by alanking** [comment]
The package is now installed at image build time to allow for service restarts. Config file for REST service is mounted to a file on the host. The configuration for the reverse-proxy is built into the nginx image. Left instructions in the README for building the packages with Docker and running with the Compose project.

File: docker/builder/Dockerfile (hunk 2 above)
Comment (trel) [reply docker/builder/Dockerfile:39]: definitely

File: docker/builder/Dockerfile (hunk 1 above)
Comment (trel) [reply docker/builder/Dockerfile:57]: definitely

**Comment by trel** [comment]
looks so good!

**Comment by korydraughn** [comment]
Agreed.

**Pull request merged**